pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java

```diff
@@ -494,50 +494,59 @@ private void processStreamEvents(MessageBatch messagesAndOffsets, long idlePipeS
-                        realtimeRowsConsumedMeter);
-                indexedMessageCount++;
-                canTakeMore = _realtimeSegment.index(transformedRow, msgMetadata);
+          Collection<GenericRow> rows = (Collection<GenericRow>) decodedRow.getValue(GenericRow.MULTIPLE_RECORDS_KEY);
+          if (rows != null) {
+            for (GenericRow row : rows) {
+              GenericRow transformedRow = _recordTransformer.transform(row);
+              if (transformedRow != null && IngestionUtils.shouldIngestRow(row)) {
+                transformedRows.add(transformedRow);
               } else {
-                realtimeRowsDroppedMeter = _serverMetrics
-                    .addMeteredTableValue(_metricKeyName, ServerMeter.INVALID_REALTIME_ROWS_DROPPED, 1,
+                realtimeRowsDroppedMeter =
+                    _serverMetrics.addMeteredTableValue(_metricKeyName, ServerMeter.INVALID_REALTIME_ROWS_DROPPED, 1,
                         realtimeRowsDroppedMeter);
               }
             }
           } else {
             GenericRow transformedRow = _recordTransformer.transform(decodedRow);
             if (transformedRow != null && IngestionUtils.shouldIngestRow(transformedRow)) {
-              realtimeRowsConsumedMeter = _serverMetrics
-                  .addMeteredTableValue(_metricKeyName, ServerMeter.REALTIME_ROWS_CONSUMED, 1,
-                      realtimeRowsConsumedMeter);
-              indexedMessageCount++;
-              canTakeMore = _realtimeSegment.index(transformedRow, msgMetadata);
+              transformedRows.add(transformedRow);
             } else {
-              realtimeRowsDroppedMeter = _serverMetrics
-                  .addMeteredTableValue(_metricKeyName, ServerMeter.INVALID_REALTIME_ROWS_DROPPED, 1,
+              realtimeRowsDroppedMeter =
+                  _serverMetrics.addMeteredTableValue(_metricKeyName, ServerMeter.INVALID_REALTIME_ROWS_DROPPED, 1,
                       realtimeRowsDroppedMeter);
             }
           }
         } catch (Exception e) {
+          _numRowsErrored++;
           String errorMessage = String.format("Caught exception while transforming the record: %s", decodedRow);
           _segmentLogger.error(errorMessage, e);
-          _numRowsErrored++;
-          _realtimeTableDataManager
-              .addSegmentError(_segmentNameStr, new SegmentErrorInfo(System.currentTimeMillis(), errorMessage, e));
+          _realtimeTableDataManager.addSegmentError(_segmentNameStr,
+              new SegmentErrorInfo(System.currentTimeMillis(), errorMessage, e));
+        }
+        for (GenericRow transformedRow : transformedRows) {
+          try {
+            canTakeMore = _realtimeSegment.index(transformedRow, msgMetadata);
+            indexedMessageCount++;
+            realtimeRowsConsumedMeter =
+                _serverMetrics.addMeteredTableValue(_metricKeyName, ServerMeter.REALTIME_ROWS_CONSUMED, 1,
```

Comment:
One consumed row can transform into multiple transformed rows, so we should bump this metric after we have consumed a row (and before we transform it) -- whether there are exceptions or not.

Reply:
This keeps the existing behavior where each transformed row bumps this metric ones. I feel we should keep the existing behavior

Reply:
Ah I see now. In the original code, we bump metrics independently. if multiple transformed rows, then once for each transformed row, otherwise once for consumed row.  Your new code will preserve the same behavior yes. thanks for the clarification.